src/Quarrel.ViewModels/Services/Gateway/GatewayService.cs

```diff
@@ -78,25 +84,32 @@ public async void InitializeGateway([NotNull] string accessToken)
 
             Gateway.SessionReplaced += Gateway_SessionReplaced;
 
-
-            await Gateway.ConnectAsync();
-
-            Messenger.Default.Register<GuildNavigateMessage>(this, async m =>
+            if(await ConnectWithRetryAsync(3))
             {
-                // TODO: Channel typing check
-                //var channelList = ServicesManager.Cache.Runtime.TryGetValue<List<Channel>>(Quarrel.Helpers.Constants.Cache.Keys.ChannelList, m.GuildId);
-                //var idList = channelList.ConvertAll(x => x.Id);
+                Messenger.Default.Register<GuildNavigateMessage>(this, async m =>
+                {
+                    // TODO: Channel typing check
+                    //var channelList = ServicesManager.Cache.Runtime.TryGetValue<List<Channel>>(Quarrel.Helpers.Constants.Cache.Keys.ChannelList, m.GuildId);
+                    //var idList = channelList.ConvertAll(x => x.Id);
 
-                List<string> idList = new List<string>();
+                    List<string> idList = new List<string>();
 
-                // Guild Sync
-                if (m.Guild.Model.Id != "DM")
-                {
-                    idList.Add(m.Guild.Model.Id);
-                }
+                    // Guild Sync
+                    if (m.Guild.Model.Id != "DM")
+                    {
+                        idList.Add(m.Guild.Model.Id);
+                    }
+
+                    await Gateway.SubscribeToGuild(idList.ToArray());
+                });
+            }
+        }
 
-                await Gateway.SubscribeToGuild(idList.ToArray());
-            });
+        public async Task<bool> ConnectWithRetryAsync(int retriesLeft)
+        {
+            if (await Gateway.ConnectAsync()) return true;
+            else if (retriesLeft > 0) { return await ConnectWithRetryAsync(--retriesLeft); }
+            else return false;
         }
```

Comment:
I think a loop would look clearer here, why the use of recursion?

Reply:
Preference, nothing else.